win32ss/user/ntuser/ghost.c

```diff
@@ -1,18 +1,168 @@
 /*
  * PROJECT:     ReactOS user32.dll
  * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
- * PURPOSE:     Ghost window handling
+ * PURPOSE:     Window ghosting feature
  * COPYRIGHT:   Copyright 2018 Katayama Hirofumi MZ (katayama.hirofumi.mz@gmail.com)
  */
 
 #include <win32k.h>
+#include "ghostwnd.h"
 
-BOOL FASTCALL IntGoGhost(PWND Window, BOOL bGo)
+#define NDEBUG
+#include <debug.h>
+
+static UNICODE_STRING GhostClass = RTL_CONSTANT_STRING(GHOSTCLASSNAME);
+static UNICODE_STRING GhostProp = RTL_CONSTANT_STRING(GHOST_PROP);
+
+BOOL FASTCALL IntIsGhostWindow(HWND hWnd)
+{
+    BOOLEAN Ret = FALSE;
+    UNICODE_STRING ClassName;
+    INT Len;
+
+    // check class name
+    RtlInitUnicodeString(&ClassName, NULL);
+    Len = NtUserGetClassName(hWnd, TRUE, &ClassName);
+    if (Len)
+    {
+        Ret = RtlEqualUnicodeString(&ClassName, &GhostClass, TRUE);
+    }
+    else
+    {
+        DPRINT1("Unable to get class name\n");
+    }
+    RtlFreeUnicodeString(&ClassName);
+
+    return Ret;
+}
+
+HWND APIENTRY NtUserGhostWindowFromHungWindow(HWND hwndHung)
+{
+    RTL_ATOM Atom;
+    PWND pHungWnd;
+    PPROPERTY Prop;
+    HWND hwndGhost;
+
+    pHungWnd = ValidateHwndNoErr(hwndHung);
+    if (!pHungWnd)
+    {
+        DPRINT("Not a window\n");
+        return NULL;
+    }
+
+    IntGetAtomFromStringOrAtom(&GhostProp, &Atom);
+
+    Prop = UserGetProp(pHungWnd, Atom, TRUE);
+
+    _SEH2_TRY
+    {
+        hwndGhost = (HWND)(Prop ? Prop->Data : NULL);
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        DPRINT1("Exception!\n");
+        hwndGhost = NULL;
+    }
+    _SEH2_END;
+
+    if (hwndGhost)
+    {
+        if (ValidateHwndNoErr(hwndGhost))
+            return hwndGhost;
+
+        DPRINT("Not a window\n");
+    }
+
+    return NULL;
+}
+
+HWND APIENTRY NtUserHungWindowFromGhostWindow(HWND hwndGhost)
 {
+    GHOST_DATA *UserData;
+    PWND pGhostWnd;
+    HWND hwndTarget;
+
+    if (!IntIsGhostWindow(hwndGhost))
+    {
+        DPRINT("Not a ghost window\n");
+        return NULL;
+    }
+
+    pGhostWnd = ValidateHwndNoErr(hwndGhost);
```

Comment:
IntIsGhostWindow necessarily already has to do this check, so it either needs to be moved earlier or it can be removed.

Reply:
OK.